lib/shared/config-manager/src/index.ts

```diff
@@ -63,28 +84,121 @@ export class EnvironmentConfigManager {
                 this.logger.debug('DevCycle initial config loaded')
             })
             .finally(() => {
-                if (this.disablePolling) {
-                    return
-                }
-                this.intervalTimeout = this.setInterval(async () => {
-                    try {
-                        await this._fetchConfig()
-                    } catch (ex) {
-                        this.logger.error((ex as Error).message)
-                    }
-                }, this.pollingIntervalMS)
+                this.startPolling(this.configPollingIntervalMS)
+                this.startSSE()
             })
     }
+
+    private startSSE(): void {
+        if (!this.enableRealtimeUpdates) return
+
+        if (!this.configSSE) {
+            this.logger.warn('No SSE configuration found')
+            return
+        }
+        if (this.sseConnection) {
+            return
+        }
+
+        const url = new URL(
+            this.configSSE.path,
+            this.configSSE.hostname,
+        ).toString()
+        this.logger.debug(`Starting SSE connection to ${url}`)
+
+        this.sseConnection = new SSEConnection(url, this.logger, {
+            onMessage: this.onSSEMessage.bind(this),
+            onOpen: () => {
+                this.logger.debug('SSE connection opened')
+                // Set config polling interval to 10 minutes
+                this.startPolling(this.sseConfigPollingIntervalMS)
+            },
+            onConnectionError: () => {
+                this.logger.debug('SSE connection error, switching to polling')
+                // reset polling interval to default
+                this.startPolling(this.configPollingIntervalMS)
+                this.stopSSE()
+            },
+        })
+    }
+
+    private onSSEMessage(message: string): void {
+        this.logger.debug(`SSE message: ${message}`)
+        try {
+            const parsedMessage = JSON.parse(message as string)
+            const messageData = JSON.parse(parsedMessage.data)
+
+            if (
+                !messageData ||
+                !(!messageData.type || messageData.type === 'refetchConfig')
+            ) {
+                return
+            }
+            if (this.configEtag && messageData.etag === this.configEtag) {
+                return
+            }
+
+            if (this.isLastModifiedHeaderOld(messageData.lastModified)) {
+                this.logger.debug(
+                    'Skipping SSE message, config last modified is newer. ',
+                )
+                return
+            }
+
+            this._fetchConfig()
+                .then(() => {
+                    this.logger.debug('Config re-fetched from SSE message')
+                })
+                .catch((e) => {
+                    this.logger.warn(
```

Comment:
is it good enough to just log that it failed and not keep retrying? before a failure wasnt as big of a deal because it would retry at the polling interval which was every minute or so. 10 minutes is a pretty long time to not receive an update.

Reply:
There is retrying in the fetch code, so if this fails we probably should go back to the shorter polling interval. But that also probably means we should then disconnect SSE, because I'm trying to keep this code somewhat simple in its state, either SSE is connected and the polling interval is long, or its disconnected and doing regular polling.

Reply:
Also just noticed we don't really have anything that's verifying that the config returned in this fetch call is equal to the one from the message or newer...

Reply:
ok based on chatting with ChatGPT I think we should switch to setting:
```
We use If-Match and If-Unmodified-Since headers to ensure that the server returns the file only if it matches the given ETag and Last-Modified values.
```

which will return a 412 error if it doesn't match: 

```
If the server returns a 412 Precondition Failed status, it means that the resource does not match the provided ETag or Last-Modified values. You can handle this accordingly to fetch the latest resource.
```

Reply:
Do we know for certain that Cloudflare supports this?

Reply:
we didn't end up going this approach and went with custom retry logic.